File: 2021/10/19/mountain_array.py
"""
https://leetcode.com/problems/valid-mountain-array/
Author: Andrew Jarombek
Date: 10/19/2021
"""

from typing import List


def valid_mountain_array(arr: List[int]) -> bool:
    if len(arr) <= 2:
        return False

    up_segment = False
    down_segment = False
    prev = arr[0]

    for i in range(1, len(arr)):
        if down_segment:
            if arr[i] >= prev:
                return False
        else:
            if arr[i] == prev:
                return False
            elif arr[i] < prev:
                down_segment = True
            else:
                up_segment = True

        prev = arr[i]

    return up_segment and down_segment

File: 2022/05/26/mountainArray.ts
/**
 * https://leetcode.com/problems/valid-mountain-array/
 * @author Andrew Jarombek
 * @since 5/26/2022
 */

function validMountainArray(arr: number[]): boolean {
    if (arr.length < 3) {
        return false;
    }

    let upwards = true;
    let up = 0;

    for (let i = 1; i < arr.length; i++) {
        if (upwards) {
            if (arr[i] < arr[i - 1]) {
                if (!up)
                    return false;

                upwards = false;
            } else if (arr[i] == arr[i - 1]) {
                return false;
            } else {
                up++
            }
        } else {
            if (arr[i] >= arr[i - 1])
                return false;
        }
    }

    return !upwards;
}